File: solutions/fsharp/protein-translation/1/ProteinTranslation.fs
module ProteinTranslation

let encode codon =
    match codon with
    | "AUG" -> Some("Methionine")
    | "UUU" | "UUC" -> Some("Phenylalanine")
    | "UUA" | "UUG" -> Some("Leucine")
    | "UCU" | "UCC" | "UCA" | "UCG" -> Some("Serine")
    | "UAU" | "UAC" -> Some("Tyrosine")
    | "UGU" | "UGC" -> Some("Cysteine")
    | "UGG" -> Some("Tryptophan")
    | "UAA" | "UAG" | "UGA" -> Some("STOP")
    | _ -> None

let rec encodeCodons acc codons =
    match codons with
    | [] -> acc
    | x :: xs ->
        match encode x with
        | Some("STOP") -> acc
        | Some(protein) -> encodeCodons (protein :: acc) xs
        | None -> acc

let proteins (rna: string) =
    rna
    |> Seq.chunkBySize 3
    |> Seq.map (System.String.Concat)
    |> Seq.toList
    |> encodeCodons []
    |> List.rev

File: solutions/haskell/protein-translation/1/src/ProteinTranslation.hs
module ProteinTranslation (proteins) where

import Data.List (find)

type Protein = String

type Codon = String

codonMapping :: [([Codon], Protein)]
codonMapping =
  [ (["AUG"], "Methionine"),
    (["UUU", "UUC"], "Phenylalanine"),
    (["UUA", "UUG"], "Leucine"),
    (["UCU", "UCC", "UCA", "UCG"], "Serine"),
    (["UAU", "UAC"], "Tyrosine"),
    (["UGU", "UGC"], "Cysteine"),
    (["UGG"], "Tryptophan"),
    (["UAA", "UAG", "UGA"], "STOP")
  ]

decodeCodon :: Codon -> Maybe Protein
decodeCodon c = snd <$> find (\(codons, _) -> c `elem` codons) codonMapping

decodeCodons :: [Protein] -> [Codon] -> [Protein]
decodeCodons acc [] = acc
decodeCodons acc (x : xs) =
  case decodeCodon x of
    Just "STOP" -> acc
    Just protein -> decodeCodons (protein : acc) xs
    Nothing -> acc

chunkBy :: Int -> [a] -> [[a]]
chunkBy _ [] = []
chunkBy n xs = take n xs : chunkBy n (drop n xs)

proteins :: String -> Maybe [Protein]
proteins rna = Just $ reverse $ decodeCodons [] $ chunkBy 3 rna
